applications/GeoMechanicsApplication/custom_elements/permeability_calculator.cpp

```diff
@@ -0,0 +1,70 @@
+// KRATOS___
+//     //   ) )
+//    //         ___      ___
+//   //  ____  //___) ) //   ) )
+//  //    / / //       //   / /
+// ((____/ / ((____   ((___/ /  MECHANICS
+//
+//  License:         geo_mechanics_application/license.txt
+//
+//  Main authors:    Richard Faasse
+//
+#include "permeability_calculator.h"
+#include "custom_retention/retention_law.h"
+#include "custom_utilities/element_utilities.hpp"
+#include "custom_utilities/transport_equation_utilities.hpp"
+#include "includes/cfd_variables.h"
+
+namespace Kratos
+{
+
+PermeabilityCalculator::PermeabilityCalculator(InputProvider InputProvider)
+    : mInputProvider{std::move(InputProvider)}
+{
+}
+
+Matrix PermeabilityCalculator::LHSContribution() { return CalculatePermeabilityMatrix(); }
+
+Vector PermeabilityCalculator::RHSContribution()
+{
+    return RHSContribution(CalculatePermeabilityMatrix());
+}
+
+std::pair<Matrix, Vector> PermeabilityCalculator::CalculateLeftAndRightHandSide()
+{
+    const auto permeability_matrix = CalculatePermeabilityMatrix();
+    return {permeability_matrix, (RHSContribution(permeability_matrix))};
+}
+
+Vector PermeabilityCalculator::RHSContribution(const Matrix& rPermeabilityMatrix) const
+{
+    return -prod(rPermeabilityMatrix, mInputProvider.GetNodalValues(WATER_PRESSURE));
+}
+
+Matrix PermeabilityCalculator::CalculatePermeabilityMatrix() const
+{
+    RetentionLaw::Parameters retention_parameters(mInputProvider.GetElementProperties());
+    const auto&              r_properties = mInputProvider.GetElementProperties();
+
+    auto r_integration_coefficients = mInputProvider.GetIntegrationCoefficients();
+
+    auto         shape_function_gradients = mInputProvider.GetShapeFunctionGradients();
```

Comment:
Also this local variable can be made `const` I reckon:
```suggestion
    const auto   shape_function_gradients = mInputProvider.GetShapeFunctionGradients();
```

Reply:
Done